crates/ide_assists/src/handlers/extract_type_alias.rs

```diff
@@ -0,0 +1,176 @@
+use syntax::{
+    ast::{self, AstNode},
+    SyntaxKind,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: extract_type_alias
+//
+// Extracts the selected type as a type alias.
+//
+// ```
+// struct S {
+//     field: $0(u8, u8, u8)$0,
+// }
+// ```
+// ->
+// ```
+// type ${0:Type} = (u8, u8, u8);
+//
+// struct S {
+//     field: ${0:Type},
+// }
+// ```
+pub(crate) fn extract_type_alias(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    if ctx.frange.range.is_empty() {
+        return None;
+    }
+
+    let node = match ctx.covering_element() {
+        syntax::NodeOrToken::Node(node) => node,
+        syntax::NodeOrToken::Token(tok) => tok.parent()?,
+    };
+    let range = node.text_range();
+    let mut type_like_node = None;
+    for node in node.ancestors() {
+        if node.text_range() != range {
+            break;
+        }
+
+        let kind = node.kind();
+        if ast::Type::can_cast(kind) || kind == SyntaxKind::TYPE_ARG {
+            type_like_node = Some(node);
+            break;
+        }
+    }
+
+    let node = type_like_node?;
+
+    let insert = ctx.find_node_at_offset::<ast::Item>()?.syntax().text_range().start();
+    let target = node.text_range();
+
+    acc.add(
+        AssistId("extract_type_alias", AssistKind::RefactorExtract),
+        "Extract type as type alias",
+        target,
+        |builder| {
+            builder.edit_file(ctx.frange.file_id);
+            match ctx.config.snippet_cap {
+                Some(cap) => {
+                    builder.replace_snippet(cap, target, "${0:Type}");
+                    builder.insert_snippet(
+                        cap,
+                        insert,
+                        format!("type ${{0:Type}} = {};\n\n", node),
+                    );
```

Comment:
:thinking: 

Three separate thoughts: 

1. the snippet's `${0}` feels like a wrong tool here: it can't express the fact that the second position should be focused (primary). 
2. Our "extract variable" works differently as well, it just foceses the new name (so that it's possible to rename it later)
3. I've noticed that type-script applies rename automatically. I *think* they just set the `command` to rename? 

So, let's implement the same logic as extract variable now, and then add something like

```diff
	/**
	 * A command this code action executes. If a code action
	 * provides an edit and a command, first the edit is
	 * executed and then the command.
	 */
	command?: Command;
```

Reply:
Oh, I see, that does seem better!

Reply:
I'll implement this part later